### PR TITLE
Order name servers by SRTT

### LINKED ISSUE #1702: Name server sorting is problematic
**Describe the bug**
The Trust-DNS resolver [sorts](https://github.com/bluejekyll/trust-dns/blob/main/crates/resolver/src/name_server/name_server_pool.rs#L185) the name servers it's been configured with, by (1) ["connection state"](https://github.com/bluejekyll/trust-dns/blob/main/crates/resolver/src/name_server/name_server.rs#L204), so that name servers that have been successfully reached are prioritized above unreachable ones; and (2) based on [stats](https://github.com/bluejekyll/trust-dns/blob/main/crates/resolver/src/name_server/name_server_stats.rs#L12) collected on how many queries to a given resolver were successful vs. failed.

However, there are a few problems with this sorting:

- Connection state (`NameServerState`) seems to be sorted backwards from what I would expect. Because `Failed = 0` and `Established = 2`, and servers are sorted in (default) increasing order, servers with which we've established communication are sorted last, while servers we've previously attempted and failed to reach are ranked first. (Maybe I'm missing something here?)
- Stats about successful/failed queries (`NameServerStats`) are collected and compared in unintuitive ways.
  - From local testing, it looks like failed queries only get recorded in stats if the name server fails to respond _and_ the overall query fails. In other words, if two of three name servers in a pool are nonresponsive, but one does eventually respond, those two name servers won't have their failure stats incremented. I need to dig deeper into why exactly this is happening.
  - Name servers with more failures are sorted later (this matches my expectations), but successes are [compared backwards](https://github.com/bluejekyll/trust-dns/blob/main/crates/resolver/src/name_server/name_server_stats.rs#L96), i.e. name servers with fewer successful queries are queried first, presumably to balance things out.

Unfortunately, when you combine the last two properties, you can get into a situation where a nonresponsive name server always gets queried before a good one as long as the overall query never fails. :( If, for example, NameServerA always drops queries and NameServerB always responds, NameServerA's failure stats won't get incremented, so you can imagine that at some point A's stats are `successes: 0 failures: 0` and B's stats are `successes: 10 failures: 0` and A will get queried first. When in practice this happens with pairs or groups of name servers, this can result in long timeouts / unexpected latency.

**To Reproduce**
Configure the resolver with multiple name servers, one or more of which is nonresponsive. Observe that over time (depending on the particulars of the network setup), the nonresponsive name server(s) get queried first.

**Expected behavior**
Here's what I think we should do:
- Short-term, we'd like name server sorting to be configurable (at least able to turn on/off)
- I need to do some more digging to fully understand how the stats and connection state are used to sort, but I think some of this sorting logic needs to be fixed
- It would be great if we could incorporate latency into `NameServerStats` (see [this TODO](https://github.com/bluejekyll/trust-dns/blob/main/crates/resolver/src/name_server/name_server_stats.rs#L15) in the code).
- Long-term, it would be great to have more fine-grained control over how this prioritization/sorting is done. As an example, it would be great to be able to turn on prioritization based on connection status (`NameServerState`), but not successes vs. failures (`NameServerStats`), or something similar.

I'm happy to take on this work, but I wanted to raise the issues that we've run into and see what folks think.

### PR DESCRIPTION
**Pull request by nhurley3**
Calculates the smoothed round-trip time (SRTT) using an exponentially weighted moving average and orders the name servers by the value. The implemented algorithm is similar to what was discussed in #1702.

Fixes #1702 (originally filed by @peterthejohnston)

### CODE REVIEW

**Comment by nhurley3** [comment]
@bluejekyll and @djc, here's a draft PR that implements a version of the algorithm mentioned in https://github.com/bluejekyll/trust-dns/issues/1702#issuecomment-1209814007. Let me know what you think. Thanks!

**Review by bluejekyll — COMMENTED** [review]
Ok, so with these changes, we’ll be ignoring the success and failures from the algorithm and only paying attention to request time. I think I’m ok with this, but it’s worth a discussion before landing.

File: crates/resolver/src/name_server/name_server.rs [hunk 1]
```diff
@@ -138,10 +138,15 @@ impl<C: DnsHandle<Error = ResolveError>, P: ConnectionProvider<Conn = C>> NameSe
         request: R,
     ) -> Result<DnsResponse, ResolveError> {
         let mut client = self.connected_mut_client().await?;
+        let now = std::time::Instant::now();
```
Comment (djc) [inline crates/resolver/src/name_server/name_server.rs:141]: Would prefer this to be just `Instant::now()`, unless there's some other type named `Instant` in scope here?

File: crates/resolver/src/name_server/name_server_stats.rs [hunk 2]
```diff
@@ -6,125 +6,295 @@
...
+   /// Prefer to use `decayed_srtt` when ordering name servers.
+   fn srtt_microseconds(&self) -> u32 {
+       self.srtt_microseconds.load(atomic::Ordering::Acquire)
+   }
+
+   /// Returns the SRTT value after applying a time based decay.
+   ///
+   /// Prefer to use this value when ordering name servers. Applying a decay
+   /// helps distribute query load and detect positive network changes.
+   fn decayed_srtt(&self) -> f64 {
```
Comment (djc) [inline crates/resolver/src/name_server/name_server_stats.rs]: Would be nice to add more discussion here of how the value decays (that is, does it decay upwards or downwards) and why.

File: crates/resolver/src/name_server/name_server_stats.rs [hunk 3]
```diff
@@ -6,125 +6,295 @@
...
+           // more uniform. A larger weight widens the window in which servers
+           // with historically lower latencies will be heavily preferred. On
+           // the other hand, a larger weight may also increase the time it
+           // takes to recover from a failure or to observe positive changes in
+           // latency.
+           srtt * compute_srtt_factor(last_update, 180)
+       })
+   }
+
+   fn update_srtt(&self, default: u32, update_fn: impl Fn(u32, Instant) -> u32) {
```
Comment (djc) [inline crates/resolver/src/name_server/name_server_stats.rs]: Maybe document the semantics of the arguments here?

File: crates/resolver/src/name_server/name_server_stats.rs [hunk 4]
```diff
@@ -6,125 +6,295 @@
...
-        other_successes: usize,
-        self_failures: usize,
-        other_failures: usize,
-    ) -> bool {
-        self_successes == other_successes && self_failures == other_failures
-    }
+   const CONNECTION_FAILURE_PENALTY: u32 = Duration::from_millis(150).as_micros() as u32;
+   const MAX_SRTT_MICROS: u32 = Duration::from_secs(5).as_micros() as u32;
+
+   pub(crate) fn new(initial_srtt: u32) -> Self {
```
[the remark this answers is not in the document]
Comment (nhurley3) [reply crates/resolver/src/name_server/name_server_stats.rs]: These are currently not test-only. The `default` implementation invokes `new`. Similarly, the constants are used to update the SRTT value.

File: crates/resolver/src/name_server/name_server_stats.rs [hunk 5]
```diff
@@ -6,125 +6,295 @@
...
+       Self {
+           srtt_microseconds: AtomicU32::new(initial_srtt),
+           last_update: Arc::new(Mutex::new(None)),
+       }
+   }
+
+   /// Records the measured `rtt` for a particular query.
+   pub(crate) fn record_rtt(&self, rtt: Duration) {
```
[the remark this answers is not in the document]
Comment (nhurley3) [reply crates/resolver/src/name_server/name_server_stats.rs]: Sorry, not sure I follow. How would this work? The scope of the returned value seems like it would just be `NameServer::inner_send`, which would mean that we'd always record a connection failure?

File: crates/resolver/src/name_server/name_server_stats.rs (hunk 2 above)
Comment (nhurley3) [reply crates/resolver/src/name_server/name_server_stats.rs]: Done.

File: crates/resolver/src/name_server/name_server_stats.rs [hunk 6]
```diff
@@ -6,125 +6,295 @@
...
-    ) -> bool {
-        self_successes == other_successes && self_failures == other_failures
-    }
+   const CONNECTION_FAILURE_PENALTY: u32 = Duration::from_millis(150).as_micros() as u32;
+   const MAX_SRTT_MICROS: u32 = Duration::from_secs(5).as_micros() as u32;
+
+   pub(crate) fn new(initial_srtt: u32) -> Self {
+       Self {
+           srtt_microseconds: AtomicU32::new(initial_srtt),
+           last_update: Arc::new(Mutex::new(None)),
```
[the remark this answers is not in the document]
Comment (nhurley3) [reply crates/resolver/src/name_server/name_server_stats.rs]: I chose to use `Instant` because it represents a monotonic clock reading, which seemed beneficial since the algorithm is time dependent. Moreover, since there is no mechanism to get the number of seconds from an `Instant`, I was forced to store the actual `Instant` value.

The alternative I considered was to use the non-monotonic `SystemTime` type and store the seconds value in an atomic. We could take this approach and just return the 1 second minimum when time drift occurs. FWIW, the algorithm used by PowerDNS is also time dependent and I don't think they use a monotonic clock.

In summary, using the `Instant` type is probably slightly more accurate, but doesn't seem like a huge deal. If you'd prefer, I'd be happy to make the switch (or perhaps you have something else in mind?).

File: crates/resolver/src/name_server/name_server_stats.rs [hunk 7]
```diff
@@ -6,125 +6,295 @@
...
+           // the other hand, a larger weight may also increase the time it
+           // takes to recover from a failure or to observe positive changes in
+           // latency.
+           srtt * compute_srtt_factor(last_update, 180)
+       })
+   }
+
+   fn update_srtt(&self, default: u32, update_fn: impl Fn(u32, Instant) -> u32) {
+       let last_update = self.last_update.lock().replace(Instant::now());
+       let _ = self.srtt_microseconds.fetch_update(atomic::Ordering::SeqCst, atomic::Ordering::SeqCst, move |cur_srtt_microseconds| {
```
[the remark this answers is not in the document]
Comment (nhurley3) [reply crates/resolver/src/name_server/name_server_stats.rs]: The issue with doing so is that `std::cmp::Ordering` is already imported.

File: crates/resolver/src/name_server/name_server_stats.rs [hunk 8]
```diff
@@ -6,125 +6,295 @@
...
-            other_successes,
-            self_failures,
-            other_failures,
-        )
-    }
+   fn eq(&self, other: &Self) -> bool {
+       let self_srtt = self.srtt_microseconds();
+       let other_srtt = other.srtt_microseconds();
+
+       self_srtt == other_srtt
```
[the remark this answers is not in the document]
Comment (nhurley3) [reply crates/resolver/src/name_server/name_server_stats.rs]: Done.

File: crates/resolver/src/name_server/name_server_stats.rs [hunk 9]
```diff
@@ -6,125 +6,295 @@
...
-        // at this point we'll go with the lesser of successes to make sure there is balance
-        self_successes.cmp(&other_successes)
-    }
+   /// Custom implementation of Ord for NameServer which incorporates the
+   /// performance of the connection into it's ranking.
+   fn cmp(&self, other: &Self) -> Ordering {
+       let self_srtt = self.decayed_srtt();
+       let other_srtt = other.decayed_srtt();
+
+       self_srtt.total_cmp(&other_srtt)
```
[the remark this answers is not in the document]
Comment (nhurley3) [reply crates/resolver/src/name_server/name_server_stats.rs]: Done.

File: crates/resolver/src/name_server/name_server_stats.rs [hunk 10]
```diff
@@ -6,125 +6,295 @@
...
+   pub(crate) fn record_connection_failure(&self) {
+       self.update_srtt(Self::CONNECTION_FAILURE_PENALTY, |cur_srtt_microseconds, _last_update| {
+           cur_srtt_microseconds.saturating_add(Self::CONNECTION_FAILURE_PENALTY)
+       });
+   }
+
+   /// Returns the raw SRTT value.
+   ///
+   /// Prefer to use `decayed_srtt` when ordering name servers.
+   fn srtt_microseconds(&self) -> u32 {
```
[the remark this answers is not in the document]
Comment (nhurley3) [reply crates/resolver/src/name_server/name_server_stats.rs]: It could yield a `Duration`, but I originally didn't choose to do so because it didn't feel like it added much value and it required slightly more verbose tests. I don't have a strong preference though and would be happy to make the change if you'd prefer.

**Comment by nhurley3** [comment]
> Ok, so with these changes, we’ll be ignoring the success and failures from the algorithm and only paying attention to request time. I think I’m ok with this, but it’s worth a discussion before landing.

This algorithm doesn't explicitly track the number of successes/failures, but they are still incorporated in some fashion. Successes now correspond to recorded latencies. Similarly, failures result in a penalty being applied to the SRTT value, which serves as a backoff.

Coupling the failure state with the SRTT value is somewhat of an arbitrary decision that simplifies things. One potentially interesting alternative could be to effectively remove failed servers from the rotation and probe them in the background (somewhat similar to what Unbound does). Of course, the tradeoff is increased complexity and additional resource cost. Happy to consider other alternatives as well. There's certainly more than 1 reasonable approach here.

File: crates/resolver/src/name_server/name_server_stats.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (djc) [reply crates/resolver/src/name_server/name_server_stats.rs]: Ah, fair.

File: crates/resolver/src/name_server/name_server_stats.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (djc) [reply crates/resolver/src/name_server/name_server_stats.rs]: Okay.

File: crates/resolver/src/name_server/name_server_stats.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (djc) [reply crates/resolver/src/name_server/name_server_stats.rs]: Makes sense!

File: crates/resolver/src/name_server/name_server_stats.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (djc) [reply crates/resolver/src/name_server/name_server_stats.rs]: Yeah, it probably doesn't make sense with the structure here. I thought it might be nice to guard against a scenario where we fail to record any SRTT for a given request.

File: crates/resolver/src/name_server/name_server_stats.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (djc) [reply crates/resolver/src/name_server/name_server_stats.rs]: I was mainly thinking about interface consistency. Right now you initialize with `new(u32)`, record with `record_rtt(Duration)` and get the value out with `srtt_microseconds() -> u32`. It seems desirable to me to use the same types for these three methods, and unless there are mitigating factors (ergonomics in non-test code, performance) I think I would prefer the more constrained/clear `Duration`.

File: crates/resolver/src/name_server/name_server_stats.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (nhurley3) [reply crates/resolver/src/name_server/name_server_stats.rs]: Thanks for explaining! Marking this as resolved since no further action is needed.

File: crates/resolver/src/name_server/name_server_stats.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (nhurley3) [reply crates/resolver/src/name_server/name_server_stats.rs]: Makes sense. Updated all 3 to use Duration.

**Review by djc — APPROVED** [review]

**Comment by bluejekyll** [comment]
Thanks for the work and getting this completed. It looks like a really good improvement. Merging.
